Refactor Events, Take 2

markandrus opened the pull request:
Here is another attempt at #364. IMO, it's easier to evaluate now, since other bugs have been fixed.

This should address the various memory leak issues, (#319, #304, and #205). The core issue is that we aren't deleting messages posted to the PeerConnection and DataChannel event loops. So, for example, if we receive a MessageEvent on a DataChannel, we never delete it, which means we leak the received buffer. Running @nazar-pc's script from #304 with these changes results in constant memory usage.

In order to solve this, I also took the opportunity to refactor our events into nicer, C++-style classes for processing via new classes, EventQueue and EventLoop. Using, e.g., `std::unique_ptr`s ensures we don't have to worry about manually `delete`-ing. Events are handled via double dispatch—basically, listing two from [Message Handling Without Dependencies](http://www.drdobbs.com/cpp/message-handling-without-dependencies/184429055). Per that article, there might be some opportunities to clean up boilerplate with this technique in the future.

- [x] Update DataChannel to use EventLoop
- [x] Update PeerConnection to use EventLoop

markandrus committed:
Refactor PeerConnection to use EventLoop

markandrus commented:
Chrome 63 and Karma on Appveyor has been really flaky recently. AFAICT, it's not a node-webrtc issue.

nazar-pc commented:
Is there anything else left to be done here? Installed some software on server and memory leaks are no fun.

markandrus commented:
@nazar-pc Well, it's super weird that the PeerConnection refactor is stalling the Appveyor tests. I wanted to get my Parallels license renewed so I could test it out on my laptop (rather than debug on Appveyor); if I can't get that setup later today, I might actually separate out the commits and land the first (there would still be leaks on the PeerConnection, though).

nazar-pc commented:
Any progress here, maybe I can help somehow?

markandrus commented:
@nazar-pc if you have a Windows machine and can test the second PR there, that'd help; otherwise, I'm still waiting on a Parallels license, so I've split off the first commit here: https://github.com/js-platform/node-webrtc/pull/390

nazar-pc commented:
I have VM with Windows 10. I don't see a list of dependencies, building process and generally I've not used Windows for anything serious in a very long time now. Could you give me short instructions?

markandrus commented:
For Windows, I think it should only require Microsoft Visual Studio Community Edition 2015 and CMake (although the Appveyor integration may document any other requirements, too). When I get my Windows VM setup again I'll see about documenting it.

markandrus committed:
Run bridge example for Chrome and Firefox separately

The bridge example can only really support on RTCPeerConnection pair at a time,
due to how offers, answers, and ICE candidates are signaled over the WebSocket.
In the long-term, we need to improve the bridge example to support multiple
RTCPeerConnections at a time. In the short-term, we'll just run the bridge
example twice.

markandrus commented:
The issue was actually with the bridge test, not the code changes.
